gadopt/approximations.py

```diff
@@ -74,20 +76,35 @@ class BoussinesqApproximation(BaseApproximation):
     All references rho, cp, alpha are constant and typically incorporated in Ra
     Viscous dissipation is neglected (Di << 1)."""
     compressible = False
+    Tbar = 0
 
-    def __init__(self, Ra, kappa=1, g=1, rho=1, alpha=1):
+    def __init__(self, Ra, *, rho=1, alpha=1, T0=0, g=1, RaB=0, delta_rho=1, kappa=1, H=0):
+        """
+        :arg Ra: Rayleigh number
+        :arg rho: Reference density
+        :arg alpha: Coefficient of thermal expansion
+        :arg T0: Reference temperature
+        :arg g: Gravitational acceleration
+        :arg RaB: Compositional Rayleigh number; product of the Rayleigh and buoyancy numbers
+        :arg delta_rho: Compositional density difference from the reference density
+        :arg kappa: Thermal diffusivity
+        :arg H: Internal heating rate
         """
-        :arg Ra:   Rayleigh number
-        :arg kappa, g, rho, alpha:  Diffusivity, gravitational acceleration, reference density and thermal expansion coefficient
-                                    Normally kept at 1 when non-dimensionalised."""
         self.Ra = ensure_constant(Ra)
-        self._kappa = ensure_constant(kappa)
-        self.g = ensure_constant(g)
         self.rho = ensure_constant(rho)
         self.alpha = ensure_constant(alpha)
+        self.T0 = ensure_constant(T0)
+        self.g = ensure_constant(g)
+        self._kappa = ensure_constant(kappa)
+        self.RaB = ensure_constant(RaB)
+        self.delta_rho = ensure_constant(delta_rho)
+        self.H = ensure_constant(H)
 
     def buoyancy(self, p, T):
-        return self.Ra * self.g * self.alpha * self.rho * T
+        return (
+            self.Ra * self.rho * self.alpha * (T - self.T0) * self.g
+            - self.RaB * self.delta_rho * self.g
+        )
```

Comment:
I've nailed the test failure on Spiegelman down to this section. I think because it has been modified to include subtractions, it goes from an expression that was identically 0 to something more like `-1 * [1.] * [0.] * [1.]` (or at least that's what it prints out for me). For some reason, this must very slightly change the order of operations or otherwise impact the accuracy of the result. I guess Kerguelen's hardware in particular is sensitive enough for this to show up, even though it passes elsewhere.

I don't know if @stephankramer knows more about the numerical implementation here to comment? I think it may end up being that we modify the expected result of this test if we verify that this is only a small relative difference, since there's not really another way around adding the extra terms to this expression.

Reply:
The one thing I do not get is that `spiegelman.py` uses `approximation = BoussinesqApproximation(0)` (line 202), so `buoyancy` necessarily returns 0, even in the updated expression. All the terms in `buoyancy` are `Constant`s except for `T`. That is confusing...

Reply:
Yes, all the terms are 0, but they're overloaded as `Constant` or whatever (some UFL subclass), so the entire expression is a `Sum` containing a few terms, which is different to when it *only* contains multiplications of terms (which reduces to identically 0). Yes, it's silly, but clearly it changes the assembly slightly!